File: Ruby/matriz.rb
class Matriz
  attr_accessor :matriz, :m, :n

  def initialize(m, n)
    @m = m
    @n = n
    @matriz = Array.new(m) { Array.new(n, 0.0) }
  end

  def self.ler_de_arquivo(caminho)
    linhas = File.readlines(caminho).map(&:strip)
    tamanho = linhas.shift.to_i
    matriz = new(tamanho, tamanho)
    tamanho.times do |i|
      matriz.matriz[i] = linhas[i].split.map(&:to_f)
    end
    matriz
  end

  def multiplicacao_escalar(escalar)
    @m.times do |i|
      @n.times do |j|
        @matriz[i][j] *= escalar
      end
    end
  end

  def soma_diagonal_superior
    soma = 0.0
    @m.times do |i|
      (i+1...@n).each do |j|
        soma += @matriz[i][j]
      end
    end
    soma
  end

  def self.multiplicacao(a, b)
    raise "Dimensões incompatíveis" unless a.n == b.m
    resultado = new(a.m, b.n)
    a.m.times do |i|
      b.n.times do |j|
        soma = 0.0
        a.n.times do |k|
          soma += a.matriz[i][k] * b.matriz[k][j]
        end
        resultado.matriz[i][j] = soma
      end
    end
    resultado
  end

  def alterar_elemento
    loop do
      print "linha: "
      m = gets.to_i
      print "coluna: "
      n = gets.to_i
      print "novo número: "
      valor = gets.to_f

      @matriz[m - 1][n - 1] = valor

      puts "Deseja ver a matriz? (1), alterar mais? (2), sair? (0)"
      opcao = gets.to_i
      imprime if opcao == 1
      break if opcao == 0
    end
  end

  def imprime
    @matriz.each do |linha|
      puts linha.map { |e| format("%.1f", e) }.join(" | ")
    end
  end

  def self.soma(a, b)
    raise "Tamanhos diferentes" unless a.m == b.m && a.n == b.n
    resultado = new(a.m, a.n)
    a.m.times do |i|
      a.n.times do |j|
        resultado.matriz[i][j] = a.matriz[i][j] + b.matriz[i][j]
      end
    end
    resultado
  end

  def self.subtrai(a, b)
    raise "Tamanhos diferentes" unless a.m == b.m && a.n == b.n
    resultado = new(a.m, a.n)
    a.m.times do |i|
      a.n.times do |j|
        resultado.matriz[i][j] = a.matriz[i][j] - b.matriz[i][j]
      end
    end
    resultado
  end

  def transposta
    resultado = Matriz.new(@n, @m)
    @m.times do |i|
      @n.times do |j|
        resultado.matriz[j][i] = @matriz[i][j]
      end
    end
    resultado
  end

  def salva_em_arquivo(nome_arquivo)
    File.open(nome_arquivo, "a") do |f|
      f.puts "#{@m} #{@n}"
      @matriz.each do |linha|
        f.puts linha.map { |e| format("%.2f", e) }.join(" ")
      end
      f.puts
    end
  end

  def determinante
    raise "Matriz não quadrada" unless @m == @n
    return @matriz[0][0] if @m == 1
    return @matriz[0][0] * @matriz[1][1] - @matriz[0][1] * @matriz[1][0] if @m == 2

    det = 0.0
    @n.times do |j|
      sub = submatriz(0, j)
      det += ((j.even? ? 1 : -1) * @matriz[0][j] * sub.determinante)
    end
    det
  end

  private

  def submatriz(linha_excluida, coluna_excluida)
    nova = Matriz.new(@m - 1, @n - 1)
    nova_linha = 0
    @matriz.each_with_index do |linha, i|
      next if i == linha_excluida
      nova_coluna = 0
      linha.each_with_index do |valor, j|
        next if j == coluna_excluida
        nova.matriz[nova_linha][nova_coluna] = valor
        nova_coluna += 1
      end
      nova_linha += 1
    end
    nova
  end
    
  def eh_nilpotente?(k)
    atual = self
    k.times do
      atual = Matriz.multiplicacao(atual, self)
      return true if atual.nula?
    end
    false
  end
  
  def nula?
    @m.times do |i|
      @n.times do |j|
        return false unless @matriz[i][j] == 0
      end
    end
    true
  end
    
  def eh_identidade?
    return false unless @m == @n
    @m.times do |i|
      @n.times do |j|
        return false if i == j && @matriz[i][j] != 1
        return false if i != j && @matriz[i][j] != 0
      end
    end
    true
  end

  def traco
    raise "Matriz não quadrada" unless @m == @n
    soma = 0.0
    @m.times { |i| soma += @matriz[i][i] }
    soma
  end
  
  def eh_simetrica?
    return false unless @m == @n
    @m.times do |i|
      @n.times do |j|
        return false unless @matriz[i][j] == @matriz[j][i]
      end
    end
    true
  end
  
  def oposta
    resultado = Matriz.new(@m, @n)
    @m.times do |i|
      @n.times do |j|
        resultado.matriz[i][j] = -@matriz[i][j]
      end
    end
    resultado
  end

  def self.produto_hadamard(a, b)
    raise "Tamanhos diferentes" unless a.m == b.m && a.n == b.n
    resultado = Matriz.new(a.m, a.n)
    a.m.times do |i|
      a.n.times do |j|
        resultado.matriz[i][j] = a.matriz[i][j] * b.matriz[i][j]
      end
    end
    resultado
  end
  

def inversa_passo_a_passo
  raise "Matriz não é quadrada!" unless @linhas == @colunas

  n = @linhas
  a = Marshal.load(Marshal.dump(@matriz))  # cópia profunda da matriz original
  identidade = Array.new(n) { |i| Array.new(n, 0.0).tap { |row| row[i] = 1.0 } }

  puts "\n--- Início do cálculo da inversa (Gauss-Jordan) ---"

  n.times do |i|
    # Verifica se o pivô é zero
    if a[i][i] == 0
      trocado = false
      (i + 1).upto(n - 1) do |j|
        if a[j][i] != 0
          a[i], a[j] = a[j], a[i]
          identidade[i], identidade[j] = identidade[j], identidade[i]
          puts "\nTrocando linha #{i} com linha #{j}:"
          trocado = true
          break
        end
      end
      raise "Matriz não possui inversa (pivô zero irremovível)" unless trocado
    end

    pivô = a[i][i]
    n.times do |j|
      a[i][j] /= pivô
      identidade[i][j] /= pivô
    end
    puts "\nLinha #{i} dividida por #{pivô}:"
    imprime_matrizes_lado_a_lado(a, identidade)

    n.times do |k|
      next if k == i
      fator = a[k][i]
      n.times do |j|
        a[k][j] -= fator * a[i][j]
        identidade[k][j] -= fator * identidade[i][j]
      end
      puts "\nEliminando elemento da linha #{k}, coluna #{i}:"
      imprime_matrizes_lado_a_lado(a, identidade)
    end
  end

  puts "\n--- Fim do cálculo da inversa ---"
  m = Matriz.new(n, n)
  m.matriz = identidade
  m
end

def imprime_matrizes_lado_a_lado(a, b)
  a.each_with_index do |linha_a, i|
    linha_b = b[i]
    puts "#{linha_a.map { |e| "%7.2f" % e }.join(" ")}  |  #{linha_b.map { |e| "%7.2f" % e }.join(" ")}"
  end
end



end

File: Py/Matriz.py
class Matriz:
    def __init__(self, m, n):
        self.m = m
        self.n = n
        self.matriz = [[0.0 for _ in range(n)] for _ in range(m)]


  @classmethod
    def ler_de_arquivo(cls, caminho):
        with open(caminho, 'r') as f:
            linhas = [linha.strip() for linha in f.readlines()]
        tamanho = int(linhas.pop(0))
        matriz = cls(tamanho, tamanho)
        for i in range(tamanho):
            matriz.matriz[i] = list(map(float, linhas[i].split()))
        return matriz



   def multiplicacao_escalar(self, escalar):
        for i in range(self.m):
            for j in range(self.n):
                self.matriz[i][j] *= escalar
                
   def soma_diagonal_superior(self):
        soma = 0.0
        for i in range(self.m):
            for j in range(i + 1, self.n):
                soma += self.matriz[i][j]
        return soma


    @staticmethod
    def multiplicacao(a, b):
        if a.n != b.m:
            raise ValueError("Dimensões incompatíveis")
        resultado = Matriz(a.m, b.n)
        for i in range(a.m):
            for j in range(b.n):
                soma = 0.0
                for k in range(a.n):
                    soma += a.matriz[i][k] * b.matriz[k][j]
                resultado.matriz[i][j] = soma
        return resultado


    def alterar_elemento(self):
        while True:
            m = int(input("linha: "))
            n = int(input("coluna: "))
            valor = float(input("novo número: "))
            self.matriz[m - 1][n - 1] = valor

            opcao = int(input("Deseja ver a matriz? (1), alterar mais? (2), sair? (0): "))
            if opcao == 1:
                self.imprime()
            if opcao == 0:
                break

    def imprime(self):
            for linha in self.matriz:
                print(" | ".join(f"{e:.1f}" for e in linha))


    @staticmethod
    def soma(a, b):
        if a.m != b.m or a.n != b.n:
            raise ValueError("Tamanhos diferentes")
        resultado = Matriz(a.m, a.n)
        for i in range(a.m):
            for j in range(a.n):
                resultado.matriz[i][j] = a.matriz[i][j] + b.matriz[i][j]
        return resultado

     @staticmethod
     def subtrai(a, b):
        if a.m != b.m or a.n != b.n:
            raise ValueError("Tamanhos diferentes")
        resultado = Matriz(a.m, a.n)
        for i in range(a.m):
            for j in range(a.n):
                resultado.matriz[i][j] = a.matriz[i][j] - b.matriz[i][j]
        return resultado

    def transposta(self):
            resultado = Matriz(self.n, self.m)
            for i in range(self.m):
                for j in range(self.n):
                    resultado.matriz[j][i] = self.matriz[i][j]
            return resultado


    def salva_em_arquivo(self, nome_arquivo):
        with open(nome_arquivo, 'a') as f:
            f.write(f"{self.m} {self.n}\n")
            for linha in self.matriz:
                f.write(" ".join(f"{e:.2f}" for e in linha) + "\n")
            f.write("\n")

    def determinante(self):
        if self.m != self.n:
            raise ValueError("Matriz não quadrada")
        if self.m == 1:
            return self.matriz[0][0]
        if self.m == 2:
            return self.matriz[0][0] * self.matriz[1][1] - self.matriz[0][1] * self.matriz[1][0]

        det = 0.0
        for j in range(self.n):
            sub = self.submatriz(0, j)
            det += ((-1) ** j) * self.matriz[0][j] * sub.determinante()
        return det

    def submatriz(self, linha_excluida, coluna_excluida):
        nova = Matriz(self.m - 1, self.n - 1)
        nova_linha = 0
        for i, linha in enumerate(self.matriz):
            if i == linha_excluida:
                continue
            nova_coluna = 0
            for j, valor in enumerate(linha):
                if j == coluna_excluida:
                    continue
                nova.matriz[nova_linha][nova_coluna] = valor
                nova_coluna += 1
            nova_linha += 1
        return nova

    def eh_nilpotente(self, k):
        atual = self
        for _ in range(k):
            atual = Matriz.multiplicacao(atual, self)
            if atual.nula():
                return True
        return False

    def nula(self):
        for i in range(self.m):
            for j in range(self.n):
                if self.matriz[i][j] != 0:
                    return False
        return True


     def eh_identidade(self):
            if self.m != self.n:
                return False
            for i in range(self.m):
                for j in range(self.n):
                    if i == j and self.matriz[i][j] != 1:
                        return False
                    if i != j and self.matriz[i][j] != 0:
                        return False
            return True

    def traco(self):
        if self.m != self.n:
            raise ValueError("Matriz não quadrada")
        return sum(self.matriz[i][i] for i in range(self.m))

    def eh_simetrica(self):
        if self.m != self.n:
            return False
        for i in range(self.m):
            for j in range(self.n):
                if self.matriz[i][j] != self.matriz[j][i]:
                    return False
        return True

    def oposta(self):
        resultado = Matriz(self.m, self.n)
        for i in range(self.m):
            for j in range(self.n):
                resultado.matriz[i][j] = -self.matriz[i][j]
        return resultado

    @staticmethod
    def produto_hadamard(a, b):
        if a.m != b.m or a.n != b.n:
            raise ValueError("Tamanhos diferentes")
        resultado = Matriz(a.m, a.n)
        for i in range(a.m):
            for j in range(a.n):
                resultado.matriz[i][j] = a.matriz[i][j] * b.matriz[i][j]
        return resultado


def inversa_gauss_jordan(self):
    if self.m != self.n:
        print("Erro: A inversa só pode ser calculada para matrizes quadradas.")
        return None

    n = self.m
    AI = [[0.0 for _ in range(2 * n)] for _ in range(n)]

    for i in range(n):
        for j in range(n):
            AI[i][j] = self.matriz[i][j]
            AI[i][j + n] = 1.0 if i == j else 0.0

    print("Matriz aumentada inicial [A | I]:")
    Matriz.imprime_matriz_estatica(AI)

    for i in range(n):
        if AI[i][i] == 0.0:
            trocou = False
            for k in range(i + 1, n):
                if AI[k][i] != 0.0:
                    AI[i], AI[k] = AI[k], AI[i]
                    trocou = True
                    print(f"\nTrocou linha {i + 1} com linha {k + 1}:")
                    Matriz.imprime_matriz_estatica(AI)
                    break
            if not trocou:
                print("Erro: Matriz singular, não possui inversa.")
                return None

        pivo = AI[i][i]
        AI[i] = [x / pivo for x in AI[i]]
        print(f"\nLinha {i + 1} dividida pelo pivô {pivo:.2f}:")
        Matriz.imprime_matriz_estatica(AI)

        for k in range(n):
            if k != i:
                fator = AI[k][i]
                AI[k] = [AI[k][j] - fator * AI[i][j] for j in range(2 * n)]
                print(f"\nEliminou elemento da linha {k + 1} usando linha {i + 1} (fator: {fator:.2f}):")
                Matriz.imprime_matriz_estatica(AI)

    inversa = Matriz(n, n)
    for i in range(n):
        for j in range(n):
            inversa.matriz[i][j] = AI[i][j + n]

    return inversa

